### ISSUE TITLE
At least some interface hints are broken in newest version.

### ISSUE THREAD

**Issue by Viissataa** [numpy/numpy#18565]
At the moment, PyCharm cannot see some functions in numpy. For example numpy.linspace is completely invisible to the IDE. 

Fallback to 1.19.5 fixes the issue.

**Comment by bashtage** [comment]
This is a PyCharm bug and should be reported on [PyCharm's bug tracker](https://youtrack.jetbrains.com/issues/PY).  They use a proprietary crawler to generate skeletons.

**Comment by east825** [comment]
Hi there! I'm a member of the PyCharm team.

The problem here is not with the skeleton generator for binary modules of numpy, but rather with .pyi stubs that the package started bundling in 1.20 ([PY-48166](https://youtrack.jetbrains.com/issue/PY-48166)). At first glance, there are a number of problems with them that greatly confuse our code insight. For instance, most of the public attributes of the top-level numpy package are annotated just as `Any` (so we no longer know that e.g. `max` is actually a function):
```python
...
mask_indices: Any
mat: Any
matrix: Any
max: Any
may_share_memory: Any
median: Any
memmap: Any
meshgrid: Any
...
```
Another notable problem is that numpy/\_\_init__.pyi defines `__all__` that contains only a small subset of the names that can actually be imported with `from numpy import *` at runtime, with a comment "This is redundant for objects defined within this file.", which, as far as I know, is not defined in any of the PEPs related to .pyi stubs (correct me if I wrong).

All in all, because we universally rely on .pyi stubs as the ultimate source of truth related to type information and resolving definitions, besides straightforward type checking, we are pretty sensitive to incomplete stubs. Are they still considered WIP? Do you need a hand or a piece of advice for them?

By the way, the skeleton generator and everything related to Numpy support, in general, are not proprietary. It's all a part of the PyCharm Community and [open-sourced](https://github.com/JetBrains/intellij-community/tree/master/python) under Apache 2.0.

**Comment by charris** [comment]
@BvB93 Thoughts? IIRC, there might be changes in main that help here. Should the be backported?

**Comment by bashtage** [comment]
I think these are in fact a WIP.  IIRC the `Any` attributes were added to stop complaints about functions not existing.  This was seen as better than getting lint about non-existing functions in downstream packages when running mypy -- but you do lose most typing until they are finished. 

Can PyCharm blacklikst using pyi files for NumPy >= 1.20 for now?

**Comment by east825** [comment]
@bashtage 
> Can PyCharm blacklikst using pyi files for NumPy >= 1.20 for now?

Yes, this is what we want to do in the short term, but at the same time, we're really interested in having thorough stubs for the package. Do you have a place where the relevant discussion happens so that we could keep in sync?

**Comment by bashtage** [comment]
As far as I know there is no comprehensive issue tracking the progress. I did the random submodule with plenty of help from @BvB93 because I am deeply familiar with it.  

It would proably be a good diea to have an issue that tracked the progress at the module and submodule level. Many of the typing tasks can be undertaken by community memebrs who are not experts in the internal workings of NumPy, but simply understand what `_ArrayLike` means.

**Comment by BvB93** [comment]
> Another notable problem is that `numpy/__init__.pyi` defines `__all__` that contains only a small subset of the names that can actually be imported with `from numpy import *` at runtime, with a comment "This is redundant for objects defined within this file.", which, as far as I know, is not defined in any of the PEPs related to .pyi stubs (correct me if I wrong).

True, hence why the respective functions were changed into explicit re-exports in https://github.com/numpy/numpy/pull/18229/commits/1dd2bd66119f73a31eb510615bfb77c193a164fa. The comment does hold true for mypy though, which is what it was initially based on. In any case, backporting the relevant commit should be trivial and, in light of your observation, a worthwhile endeavour IMO.

> For instance, most of the public attributes of the top-level numpy package are annotated just as `Any` (so we no longer know that e.g. `max` is actually a function)

This is very much a consequence of the typing of numpy being an ongoing effort, the alternative being a module-level `__getattr__` (which provides even less type safety). We're getting more into the teritory of enhancements here, but how would replacing `Any` with either `Callable[..., Any]` or an explicit function impact PyCharm?

**Comment by BvB93** [comment]
FYI the https://github.com/numpy/numpy/commit/1dd2bd66119f73a31eb510615bfb77c193a164fa backport is currently up at https://github.com/numpy/numpy/pull/18764. 
Based on current discussion I don't think it's enough to close the issue, but feel free to correct me here.

**Comment by east825** [comment]
> This is very much a consequence of the typing of numpy being an ongoing effort, the alternative being a module-level __getattr__ (which provides even less type safety). We're getting more into the teritory of enhancements here, but how would replacing Any with either Callable[..., Any] or an explicit function impact PyCharm?

Having explicit functions with all the type information about their signatures would obviously be the best, but, I guess that we could even come up with some heuristic: for instance, if something in the top-level numpy stub is still declared as `Any`, we could fall back to our existing resolve and type inference to deduce whatever we can. However, at the moment the problem was that we couldn't even automatically map these immediate declarations in numpy/\_\_init__.pyi to the corresponding definitions in inner pure Python modules and generated skeletons. My guess was that it might have been caused by some intermediate stubs for subpackages and submodules of numpy like the empty stub `core/__init__.pyi`, but I need to investigate it a bit more thoroughly.

**Comment by BvB93** [comment]
Just to give an update:
* With https://github.com/numpy/numpy/pull/18764 merged I expect; this changes should be included in the next 1.20.3 release. I believe that this should fix the immediate issue of not being able to see certain numpy functions.
* Work has been done in improving the `Any`-based placeholder annotations in the main numpy namespace (see https://github.com/numpy/numpy/issues/18838) and the plan is to do this for all remaining sub-modules before the next 1.21 release.

**Comment by BvB93** [comment]
Hi all, with the release of numpy 1.20.3 earlier today the fix for this issue is now live.
If there are no further objections then I'll close this issue.
